### PR TITLE
feat: add error handling and validation to form builder, ArrayControl style updates

### PR DESCRIPTION
**Pull request by karrui**
### TL;DR

Improved form validation and error handling in the editing experience, enhancing user feedback and preventing saves with invalid data.

Also updated ArrayControl with UI improvements

### What changed?

- Added an ErrorProvider component to manage form errors across the application
- Updated BaseBlock, DraggableBlock, and DraggableDrawerButton components with new styling and error states
- Modified ComplexEditorStateDrawer, HeroEditorDrawer, and MetadataEditorStateDrawer to use the new ErrorProvider
- Implemented a new SaveButton component that disables when errors are present
- Updated FormBuilder to use the ErrorProvider for setting errors
- Enhanced JsonFormsArrayControl with better error handling and UI improvements

### How to test?

1. Navigate to the page editor
2. Edit various components and intentionally introduce errors (e.g., clear required fields)
3. Verify that error states are displayed correctly in the UI
4. Attempt to save changes and confirm that the save button is disabled when errors are present
5. Fix the errors and verify that the save button becomes enabled

### Why make this change?

This change improves the user experience by providing immediate feedback on form errors and preventing users from saving invalid data. It enhances the overall robustness of the editing experience and helps maintain data integrity.

### CODE REVIEW

**Comment by karrui** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/608?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#608** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/608?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#603** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/603?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#599** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/599?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#592** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/592?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @karrui and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/DraggableDrawerButton.tsx [hunk 1]
```diff
@@ -6,81 +6,117 @@ import type {
...
             <Icon as={BiGridVertical} fontSize="1.5rem" color="slate.300" />
           </Box>
           <Box
+            layerStyle="focusRing"
             as="button"
             onClick={() => setSelectedIndex(index)}
             pl="0.25rem"
             pr="1rem"
-            py="0.75rem"
+            py={isError ? "0.75rem" : "1rem"}
```
Comment (dcshzj) [inline apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/DraggableDrawerButton.tsx:97]: issue: this still feels imbalanced

image

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/DraggableDrawerButton.tsx [hunk 2]
```diff
@@ -6,81 +6,117 @@ import type {
...
+              width="6px"
+              mr="-6px"
+            />
+          )}
           <Box
             cursor="grab"
             pl="0.5rem"
             pr="0.25rem"
-            py="0.75rem"
+            py="1.25rem"
```
Comment (dcshzj) [inline apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/DraggableDrawerButton.tsx:85]: issue: the drag handle doesn't seem to be middle aligned

image

**Review by dcshzj — APPROVED** [review]

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/DraggableDrawerButton.tsx (hunk 1 above)
Comment (karrui) [reply apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/DraggableDrawerButton.tsx:97]: fixed

**Pull request merged**